design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,482 @@
+are easily discoverable, and allows the OAM runtime to inject parameters into
+any field of the embedded workload without introducing a [“fromParams()” inline
+function DSL]. This approach also allows the published parameters to omit their
+type - the type can be inferred from the type of the workload fields a parameter
+corresponds to.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: WorkloadDefinition
+metadata:
+  name: rediscluster.cache.crossplane.io
+spec:
+  definitionRef:
+    name: rediscluster.cache.crossplane.io
+```
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: Component
+metadata:
+  name: example-cache
+spec:
+  workload:
+    apiVersion: cache.crossplane.io/v1alpha1
+    kind: RedisCluster
+    spec:
+      engineVersion: "3.2"
+  parameters:
+  - name: instanceName
+    required: true
+    fieldPaths:
+    - ".metadata.name"
+  - name: secret
+    required: true
+    fieldPaths:
+    - ".spec.writeConnectionSecretToRef.name"
+  - name: engineVersion
+    required: false
+    fieldPaths:
+    - ".spec.engineVersion"
+```
+
+The above WorkloadDefinition and Component leverages the existing Crossplane
+RedisCluster resource claim. Crossplane as a tentative OAM runtime need not
+define an extended workload type; it need only create a WorkloadDefinition
+pointing to [the existing CustomResourceDefinition] of the RedisCluster resource
+claim.
+
+The Component design enables OAM runtimes to break the task of reconciling the
+desired state of an application with the actual state of the world into two
+distinct stages:
+
+1. Render workloads. The OAM runtime iterates over the Components referenced by
+   an ApplicationConfiguration, instantiating a distinct workload resource for
+   each component. In the above examples this would involve creating a distinct
+   Server resource, and a distinct RedisCluster resource.
+2. Reconcile workloads. By producing valid instances of existing Kubernetes
+   custom resources OAM is able to leverage existing controllers. For example
+   the existing Crossplane RedisClaim controller(s) would be responsible for
+   reconciling the RedisCluster custom resource produced during the render
+   phase.
+
+Note that this two phase reconciliation and the production of distinct workload
+resources would not be required by the OAM specification. Runtimes could choose
+to iterate over the referenced components, render them in memory, and act upon
+the rendered components directly.
+
+### Traits
+
+Like workloads, each trait is of a distinct kind. A TraitDefinition defines the
+schema of each kind of trait, and limits what workload kinds it applies to:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: TraitDefinition
+metadata:
+  name: manualscalertrait.core.oam.dev
+spec:
+  appliesToWorkloads:
+    - server.core.oam.dev/v1alpha1
+    - worker.core.oam.dev/v1alpha1
+    - task.core.oam.dev/v1alpha1
+  definitionRef:
+    name: manualscalertrait.core.oam.dev
```

Comment:
```suggestion
    indexName: v1alpha1.manualscalertrait.core.oam.dev
```

Reply:
Or
```
name: manualscalertrait.core.oam.dev
version: v1alpha1
```

Reply:
I'd suggest we do not distribute discussion into multiple issues. 

Also, this LGTM because `version` could be an optional field.

Reply:
I'd like to ensure this remains a strict subset of the `ObjectReference` type in accordance with API conventions per https://github.com/kubernetes/community/blob/1d6d12/contributors/devel/sig-architecture/api-conventions.md#references-to-related-objects.

Reply:
I wonder if there is any particular reason to adhere to `ObjectReference` type?  Is the controller translate our definitionRef to it?

Reply:
@ryanzhang-oss It's an API consistency constraint, not a technical one. So the reason is mostly just that Kubernetes API conventions bless this pattern as the standard one, and thus it's the least surprising option for folks interacting with our very-Kubernetes-like API.